File: code/399.evaluate-division.php
<?php
/*
 * @lc app=leetcode id=399 lang=php
 *
 * [399] Evaluate Division
 */

// @lc code=start
class Solution {

    /**
     * @param String[][] $equations
     * @param Float[] $values
     * @param String[][] $queries
     * @return Float[]
     */
    function calcEquation($equations, $values, $queries) {
        $map = [];
        foreach ($equations as $index => $equation) {
            $map[$equation[0]][$equation[1]] = $values[$index];
            $map[$equation[1]][$equation[0]] = 1 / $values[$index];
        }

        $answers = [];
        foreach ($queries as $query) {
            $answer = $this->helper($query, $map, []);
            $answers[] = $answer > 0 ? $answer : -1;
        }

        return $answers;
    }

    function helper($query, &$map, $visited) {
        $a = $query[0];
        $b = $query[1];

        $visited[$a] = true;

        if (!isset($map[$a])) return -1;
        if (!isset($map[$b])) return -1;
        if ($a === $b) return 1;
        if (isset($map[$a][$b])) {
            return $map[$a][$b];
        }

        $answers = [];
        foreach ($map[$a] as $a1 => $value) {
            if (!isset($visited[$a1])) {
                $answers[] = $value * $this->helper([$a1, $b], $map, $visited);
            }
        }

        if (empty($answers)) return -1;

        $answer = max($answers);
        if ($answer > 0) {
            $map[$a][$b] = $answer;
        }

        return $answer;
    }
}
// @lc code=end

File: code/399.evaluate-division.go
package code

/*
 * @lc app=leetcode id=399 lang=golang
 *
 * [399] Evaluate Division
 */

// @lc code=start
func calcEquation(equations [][]string, values []float64, queries [][]string) []float64 {
	graph := make(map[string]map[string]float64)
	for i, equation := range equations {
		if graph[equation[0]] == nil {
			graph[equation[0]] = make(map[string]float64)
		}
		if graph[equation[1]] == nil {
			graph[equation[1]] = make(map[string]float64)
		}
		graph[equation[0]][equation[1]] = values[i]
		graph[equation[1]][equation[0]] = 1 / values[i]
	}

	results := make([]float64, 0, len(queries))
	for _, query := range queries {
		visited := make(map[string]bool)
		result := rec399(query[0], query[1], graph, visited)
		results = append(results, result)
	}

	return results
}

func rec399(a, b string, graph map[string]map[string]float64, visited map[string]bool) float64 {
	if graph[a] == nil {
		return -1.0
	}
	if graph[b] == nil {
		return -1.0
	}
	if result, ok := graph[a][b]; ok {
		return result
	}

	for a1, _ := range graph[a] {
		if !visited[a1] {
			visited[a1] = true
			tmp := rec399(a1, b, graph, visited)
			if tmp == -1.0 {
				continue
			} else {
				return tmp * graph[a][a1]
			}
		}
	}

	return -1.0
}

// @lc code=end
